File: server/VideoService/src/main/scala/APIs/VideoService/UploadVideoMessage.scala
package APIs.VideoService


import Common.API.API
import Common.Serialize.CustomColumnTypes.decodeDateTime
import Common.Serialize.CustomColumnTypes.encodeDateTime
import Common.Serialize.JacksonSerializeUtils
import Global.ServiceCenter.VideoServiceCode
import com.fasterxml.jackson.core.`type`.TypeReference
import io.circe.Decoder
import io.circe.Encoder
import io.circe.Json
import io.circe.generic.semiauto.deriveDecoder
import io.circe.generic.semiauto.deriveEncoder
import io.circe.parser.*
import io.circe.syntax.*
import java.util.UUID
import org.joda.time.DateTime
import scala.util.Try

/**
 * UploadVideoMessage
 * desc: 根据用户Token校验身份后，通过给定的入参上传视频信息，并生成videoID存储到视频表。
 * @param token: String (用户身份校验的Token)
 * @param title: String (视频标题)
 * @param description: String (视频简介)
 * @param tag: String (视频标签列表)
 * @return videoID: Int (生成的视频ID)
 */

case class UploadVideoMessage(
  token: String,
  title: String,
  description: String,
  tag: List[String]
) extends API[Int](VideoServiceCode)

case object UploadVideoMessage{

  private val circeEncoder: Encoder[UploadVideoMessage] = deriveEncoder
  private val circeDecoder: Decoder[UploadVideoMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[UploadVideoMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[UploadVideoMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[UploadVideoMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given uploadVideoMessageEncoder: Encoder[UploadVideoMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given uploadVideoMessageDecoder: Decoder[UploadVideoMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }

}

File: client/src/Plugins/VideoService/APIs/UploadVideoMessage.ts
/**
 * UploadVideoMessage
 * desc: 根据用户Token校验身份后，通过给定的入参上传视频信息，并生成videoID存储到视频表。
 * @param token: String (用户身份校验的Token)
 * @param title: String (视频标题)
 * @param description: String (视频简介)
 * @param tag: String (视频标签列表)
 * @return videoID: Int (生成的视频ID)
 */
import { TongWenMessage } from 'Plugins/TongWenAPI/TongWenMessage'
import { ServerAddr } from '../../../server-config';



export class UploadVideoMessage extends TongWenMessage {
    constructor(
        public  token: string,
        public  title: string,
        public  description: string,
        public  tag: string[],
    ) {
        super()
    }
    getAddress(): string {
        return ServerAddr["Video"]
    }
}

